graph/simple_graph.py

```diff
@@ -0,0 +1,65 @@
+class Graph(object):
+    """Implements a graph data structure"""
+
+    def __init__(self):
+        self.graph_dict = {}
+
+    def nodes(self):
+        """return a list of all nodes in the graph"""
+        return self.graph_dict.keys()
+
+    def edges(self):
+        """return a list of all edges in the graph"""
+
+        return [(key, node) for key, value in
+                self.graph_dict.iteritems() for node in value]
+
+    def add_node(self, node):
+        """add a new node 'n' to the graph"""
+        self.graph_dict.setdefault(node, [])
+
+    def add_edge(self, node_1, node_2):
+        """add a new edge to the graph connecting 'n1' and 'n2', if either n1
+        or n2 are not already present in the graph, they are added.
+        """
+        self.add_node(node_1)
+        self.add_node(node_2)
+        self.graph_dict[node_1].append(node_2)
+
+    def del_node(self, node):
+        """delete the node 'n' from the graph"""
+        try:
+            del self.graph_dict[node]
+            for val_list in self.graph_dict.values():
+                if node in val_list:
+                    val_list.remove(node)
+        except KeyError:
+            raise KeyError("Node not found")
```

Comment:
It might be preferable to simply let the existing `KeyError` bubble up rather than replacing it with a new one.  Especially if you can mutate the message on the existing one and then re-raise it:

```python
except KeyError as e:
    e.msg = "Node Not Found"  # this is a blind guess at how one might do this
    raise
```

The problem with your implementation is that it changes the location of the error from 32 (where the error actually happened) to 37 (where the new error is raised).  This might be confusing to a user, since the locus of the problem is different than reported.